app/src/main/java/org/schabi/newpipe/player/datasource/RefreshableHlsHttpDataSource.java

```diff
@@ -0,0 +1,248 @@
+import org.schabi.newpipe.extractor.stream.RefreshableStream;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class RefreshableHlsHttpDataSource extends LoggingHttpDataSource {
+
+    private final String TAG =
+            RefreshableHlsHttpDataSource.class.getSimpleName() + "@" + hashCode();
+    private final RefreshableStream refreshableStream;
+    private final String originalPlaylistUrl;
+    private final Map<String, String> chunkUrlMap = new LinkedHashMap<>();
+    private boolean isError;
+
+    public RefreshableHlsHttpDataSource(final RefreshableStream refreshableStream) {
+        this.refreshableStream = refreshableStream;
+        originalPlaylistUrl = refreshableStream.initialUrl();
+    }
+
+    @SuppressWarnings("checkstyle:LineLength")
+    public RefreshableHlsHttpDataSource(final RefreshableStream refreshableStream,
+                                        @Nullable final String userAgent,
+                                        final int connectTimeoutMillis,
+                                        final int readTimeoutMillis,
+                                        final boolean allowCrossProtocolRedirects,
+                                        @Nullable final RequestProperties defaultRequestProperties) {
+        super(userAgent,
+              connectTimeoutMillis,
+              readTimeoutMillis,
+              allowCrossProtocolRedirects,
+              defaultRequestProperties);
+        this.refreshableStream = refreshableStream;
+        originalPlaylistUrl = refreshableStream.initialUrl();
+    }
+
+    @Override
+    public long open(final DataSpec dataSpec) throws HttpDataSourceException {
+        final var url = dataSpec.uri.toString();
+        if (DEBUG) {
+            Log.d(TAG, "called open(" + url + ")");
+        }
+
+        if (!url.contains(refreshableStream.playlistId())) {
+            // TODO: throw error or no?
+            if (DEBUG) {
+                Log.e(TAG, "Playlist id does not match");
+            }
+        }
+        return chunkUrlMap.isEmpty()
+                ? openInternal(dataSpec)
+                : openInternal(getUpdatedDataSpec(dataSpec));
+    }
+
+    private long openInternal(final DataSpec dataSpec) throws HttpDataSourceException {
+        try {
+            final var bytesToRead = super.open(dataSpec);
+            if (DEBUG) {
+                Log.d(TAG, "Bytes to read: " + bytesToRead);
+            }
+            isError = false; // if we got to this line there was no error
+            return bytesToRead;
+        } catch (final InvalidResponseCodeException e) {
+            // TODO: This assumes SoundCloud returning 403 when playlist expires
+            //  If we need to refresh playlists for other services at a later date then
+            //  need to generalize this class

```

Comment:
This can be generalized later. We might need to take a look into it anyway because DataSpec is deprecated. But it is good to keep the comment / TODO for later.